File: ai-analysis-mcp/test/mockOllama.ts
import { createServer, type Server } from "node:http";

/**
 * Emulates just enough of Ollama's REST API (/api/chat, /api/tags) to drive
 * the client and analysis code end-to-end without a real Ollama install.
 * Responses are scripted by marker strings in the user message.
 */
export const MOCK_DIGEST = "sha256:abc123def456";

function scriptedContent(userMessage: string): string {
  if (userMessage.includes("TRIGGER_MALFORMED")) return "this is not valid json {{{";

  if (userMessage.includes("TRIGGER_HALLUCINATE")) {
    return JSON.stringify({
      narrative: "The account was compromised via a phished credential.",
      confidence: "high",
      evidence_refs: ["EV-9999-DOES-NOT-EXIST"],
      insufficient_evidence: false,
    });
  }

  if (userMessage.includes("TRIGGER_NO_EVIDENCE")) {
    return JSON.stringify({
      narrative: "This account is definitely compromised.",
      confidence: "high",
      evidence_refs: [],
      insufficient_evidence: false,
    });
  }

  if (userMessage.includes("TRIGGER_INSUFFICIENT")) {
    return JSON.stringify({
      narrative: "The evidence provided does not support a firm conclusion.",
      confidence: "low",
      evidence_refs: [],
      insufficient_evidence: true,
    });
  }

  if (userMessage.includes("TRIGGER_CITE_FINDING_ID")) {
    // A real model does this: it cites the finding id alongside the evidence,
    // because the case-summary prompt hands it whole findings.
    return JSON.stringify({
      narrative: "One account shows impossible travel and a suspicious mail rule.",
      confidence: "medium",
      evidence_refs: ["F-1006", "EV-3381"],
      insufficient_evidence: false,
    });
  }

  if (userMessage.includes("TRIGGER_INVENTED_ID")) {
    return JSON.stringify({
      narrative: "A confident summary.",
      confidence: "high",
      evidence_refs: ["F-9999-NOT-REAL"],
      insufficient_evidence: false,
    });
  }

  if (userMessage.includes("TRIGGER_BAD_CONFIDENCE")) {
    return JSON.stringify({
      narrative: "A grounded narrative.",
      confidence: "very high",
      evidence_refs: ["EV-3381"],
      insufficient_evidence: false,
    });
  }

  if (userMessage.includes("TRIGGER_FALSE_CONTENT_CLAIM")) {
    return JSON.stringify({
      method: "content_analysis",
      sensitivity_flags: ["financial_banking_details"],
      narrative: "The message body contains full bank account details.",
      confidence: "high",
    });
  }

  if (userMessage.includes("TRIGGER_BAD_FLAG")) {
    return JSON.stringify({
      method: "subject_line_fallback",
      sensitivity_flags: ["nuclear_launch_codes"],
      narrative: "Subject line only.",
      confidence: "low",
    });
  }

  if (userMessage.includes("TRIGGER_SUBJECT_FALLBACK")) {
    return JSON.stringify({
      method: "subject_line_fallback",
      sensitivity_flags: ["invoice_or_payment_request"],
      narrative: "Subject line only; suggests a payment request. Limited assessment.",
      confidence: "low",
    });
  }

  if (userMessage.includes("Subject:")) {
    return JSON.stringify({
      method: "content_analysis",
      sensitivity_flags: ["invoice_or_payment_request"],
      narrative: "The body discusses an invoice payment.",
      confidence: "medium",
    });
  }

  return JSON.stringify({
    narrative: "Two sign-ins minutes apart from different countries indicate session theft.",
    confidence: "high",
    evidence_refs: ["EV-3381", "EV-3382"],
    insufficient_evidence: false,
  });
}

export interface MockOllamaOptions {
  /** Respond to every request with this HTTP status and a JSON error body,
   *  so the client's response.ok handling can be exercised. */
  status?: number;
}

export class MockOllamaServer {
  private server: Server | undefined;
  private readonly options: MockOllamaOptions;
  url = "";

  constructor(options: MockOllamaOptions = {}) {
    this.options = options;
  }

  async start(): Promise<string> {
    this.server = createServer((req, res) => {
      if (this.options.status !== undefined) {
        res.writeHead(this.options.status, { "content-type": "application/json" });
        res.end(JSON.stringify({ error: "upstream failure" }));
        return;
      }
      if (req.method === "GET" && req.url === "/api/tags") {
        res.writeHead(200, { "content-type": "application/json" });
        res.end(JSON.stringify({
          models: [{ model: "llama3.1:latest", name: "llama3.1:latest", digest: MOCK_DIGEST }],
        }));
        return;
      }
      if (req.method === "POST" && req.url === "/api/chat") {
        let raw = "";
        req.on("data", (c) => (raw += c));
        req.on("end", () => {
          const body = JSON.parse(raw || "{}") as { messages?: Array<{ role: string; content: string }> };
          const user = body.messages?.find((m) => m.role === "user")?.content ?? "";
          res.writeHead(200, { "content-type": "application/json" });
          res.end(JSON.stringify({
            model: "llama3.1",
            created_at: "2026-08-20T00:00:00Z",
            message: { role: "assistant", content: scriptedContent(user) },
            done: true,
          }));
        });
        return;
      }
      res.writeHead(404, { "content-type": "application/json" });
      res.end(JSON.stringify({ error: "not found" }));
    });

    await new Promise<void>((done) => this.server!.listen(0, "127.0.0.1", done));
    const address = this.server!.address();
    if (address === null || typeof address === "string") throw new Error("no port");
    this.url = `http://127.0.0.1:${address.port}`;
    return this.url;
  }

  async stop(): Promise<void> {
    if (this.server) await new Promise<void>((done) => this.server!.close(() => done()));
  }
}

File: ai-analysis-mcp/tests/mock_ollama.py
"""A minimal local HTTP server that mimics just enough of Ollama's REST API
(/api/chat, /api/tags) to test our client and analysis code end-to-end
without a real Ollama installation. Response content is controlled by
special marker strings in the request's user message, so tests can drive
specific scenarios (happy path, hallucinated evidence, malformed JSON,
server error).
"""

from __future__ import annotations

import json
import threading
from http.server import BaseHTTPRequestHandler, HTTPServer


class MockOllamaHandler(BaseHTTPRequestHandler):
    def log_message(self, *args):  # silence default request logging
        pass

    def do_GET(self):
        if self.path == "/api/tags":
            self._respond(200, {
                "models": [{"model": "llama3.1:latest", "digest": "sha256:abc123def456", "name": "llama3.1:latest"}]
            })
        else:
            self._respond(404, {"error": "not found"})

    def do_POST(self):
        if self.path != "/api/chat":
            self._respond(404, {"error": "not found"})
            return

        length = int(self.headers.get("Content-Length", 0))
        body = json.loads(self.rfile.read(length) or b"{}")
        user_msg = next((m["content"] for m in body.get("messages", []) if m["role"] == "user"), "")

        content = self._scripted_content(user_msg)
        self._respond(200, {
            "model": body.get("model", "llama3.1"),
            "created_at": "2026-08-14T00:00:00Z",
            "message": {"role": "assistant", "content": content},
            "done": True,
        })

    def _scripted_content(self, user_msg: str) -> str:
        if "TRIGGER_MALFORMED" in user_msg:
            return "this is not valid json {{{"
        if "TRIGGER_HALLUCINATE" in user_msg:
            return json.dumps({
                "narrative": "The account was compromised via a phished credential.",
                "confidence": "high",
                "evidence_refs": ["EV-DOES-NOT-EXIST"],
                "insufficient_evidence": False,
            })
        if "TRIGGER_NO_EVIDENCE" in user_msg:
            return json.dumps({
                "narrative": "This looks suspicious.",
                "confidence": "medium",
                "evidence_refs": [],
                "insufficient_evidence": False,
            })
        if "TRIGGER_INSUFFICIENT" in user_msg:
            return json.dumps({
                "narrative": "The evidence provided does not clearly indicate compromise.",
                "confidence": "low",
                "evidence_refs": [],
                "insufficient_evidence": True,
            })
        if "TRIGGER_SUBJECT_FALLBACK" in user_msg:
            return json.dumps({
                "method": "subject_line_fallback",
                "sensitivity_flags": ["invoice_or_payment_request"],
                "narrative": "Subject line suggests a payment-related request; body was not available for review.",
                "confidence": "low",
            })
        # default: well-formed, grounded response referencing whatever
        # evidence ids actually appear in the prompt (evidence ids are
        # conventionally prefixed "EV-" in this test suite; this avoids
        # accidentally also capturing a finding's own "id" field)
        import re
        ids = re.findall(r'"id":\s*"(EV-[^"]+)"', user_msg)
        return json.dumps({
            "narrative": "The sign-in pattern and mailbox rule together indicate the account was compromised and used to redirect financial correspondence.",
            "confidence": "high",
            "evidence_refs": ids,
            "insufficient_evidence": False,
        })

    def _respond(self, status: int, payload: dict):
        data = json.dumps(payload).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)


class MockOllamaServer:
    """Context manager: starts the mock server on a free local port in a
    background thread, yields its base URL."""

    def __init__(self):
        self._httpd = HTTPServer(("127.0.0.1", 0), MockOllamaHandler)
        self._thread = threading.Thread(target=self._httpd.serve_forever, daemon=True)

    @property
    def url(self) -> str:
        port = self._httpd.server_address[1]
        return f"http://127.0.0.1:{port}"

    def __enter__(self):
        self._thread.start()
        return self

    def __exit__(self, *exc):
        self._httpd.shutdown()
        self._httpd.server_close()
